azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/sync_clients.py

```diff
@@ -151,6 +154,52 @@ def callback():
         self._transport.send_event(message, callback=callback)
         send_complete.wait()
 
+    def receive_method(self, method_name=None, block=True, timeout=None):
+        """Receive a method call via the Azure IoT Hub or Azure IoT Edge Hub.
+
+        :param str method_name: Optionally provide the name of the method to receive calls for.
+        If this parameter is not given, all methods not already being specifically targeted by
+        a different call to receive_method will be received.
+        :param bool block: Indicates if the operation should block until a message is received.
+        Default True.
+        :param int timeout: Optionally provide a number of seconds until blocking times out.
+
+        :raises: InboxEmpty if timeout occurs on a blocking operation.
+        :raises: InboxEmpty if no message is available on a non-blocking operation.
+
+        :returns: MethodCall object representing the received method call.
+        """
+        if not self._transport.feature_enabled[constant.METHODS]:
+            self._enable_feature(constant.METHODS)
+
+        if method_name:
+            method_inbox = self._inbox_manager.get_named_method_call_inbox(method_name)
+        else:
+            method_inbox = self._inbox_manager.get_generic_method_call_inbox()
+
+        logger.info("Waiting for method call...")
+        method_call = method_inbox.get(block=block, timeout=timeout)
+        logger.info("Received method call")
+        return method_call
+
+    def send_method_response(self, method, result, status):
```

Comment:
I'm curious why this is on the client instead of putting it on the MethodCall object.  We've debated this a few times, and I always thought the MethodCall was a better place, but I don't know if those debates were ever finalized.

Reply:
They haven't really been explicitly finalized. This is as good a place as any to continue the discussion.

Some reasoning I chose to do it this way (all of which, by pure coincidence, begin with the letter 'S'):

- Simplicity of implementation. Passing callbacks through to MethodCall objects, as with the early version of Message objects (before the ack feature was scrapped), adds a lot of code, specifically code that involves passing around method pointers through multiple structures (Transport -> Client -> Inbox Manager -> MethodCall) in order to get transport functionality onto an external, user facing object like MethodCall. It's very ugly, and not particularly simple to follow along with.
- Symmetry. Receiving methods on the client, but sending the response from the MethodCall is kind of weird. 
- Semantics. Responding to a method call, is a client operation. It is the client that responds to the method, not the method itself. So, I think it's correct for the response to be done via the client. It makes clear that it is an I/O operation of the client, and that I/O is occurring.
- Separation of concern. This object represents the data of a method call. Combining that with actual call functionality creates an object with multiple purposes, which, I'm told, is not an ideal design. Objects should have one purpose.

---
In reply to: [266090705](https://github.com/Azure/azure-iot-sdk-python-preview/pull/61#discussion_r266090705) [](ancestors = 266090705)

Reply:
this is the whole proxy/dto vs "smart" objects that we never really solved. historically, IoT SDKs lean towards having everythign on the client (that acts as "proxy" and having dumb DTO that only contain data. I think it could be a part of the user study feedback we want to ask. this has consequences:
- if we put everything on the client:
  - we also put accept/reject/abandon (C2D ack methods) on the client
  - we also put reported property updates of the twin of the client
- if we make "smarter" objects:
  - accept/reject/abandon have to live on the Message object
  - update reported properties have to live on the twin object
  - file upload notification have to live on a file upload request object

and so on. I for one could go either way, i just want consistency.

Another element for this debate:  if we go with a "pipeline" model for the client, then the proxy/dto thing might be easier to reason about. I would suggest talking to Cris Pop about this.

---
In reply to: [266138632](https://github.com/Azure/azure-iot-sdk-python-preview/pull/61#discussion_r266138632) [](ancestors = 266138632,266090705)

Reply:
Yeah, I think for now the simplest thing to do is leave it on the client as is, and we can address how we want to deal with this in the (near) future. At the end of the day, I'm also on board for either approach as long as it's consistent.

---
In reply to: [266185870](https://github.com/Azure/azure-iot-sdk-python-preview/pull/61#discussion_r266185870) [](ancestors = 266185870,266138632,266090705)